backend/hidrive/helpers.go

```diff
@@ -0,0 +1,902 @@
+	SortBySize                 SortByField = "size"
+	SortByNameDescending       SortByField = SortByField(descendingSort) + SortByName
+	SortByModTimeDescending    SortByField = SortByField(descendingSort) + SortByModTime
+	SortByObjectTypeDescending SortByField = SortByField(descendingSort) + SortByObjectType
+	SortBySizeDescending       SortByField = SortByField(descendingSort) + SortBySize
+)
+
+var (
+	// Unsorted disables sorting and can therefore not be combined with other values.
+	Unsorted = []SortByField{"none"}
+	// DefaultSorted does not specify how to sort and
+	// therefore implies the default sort order.
+	DefaultSorted = []SortByField{}
+)
+
+// CopyOrMoveOperationType represents the possible types of copy- and move-operations.
+type CopyOrMoveOperationType int
+
+// possible values for CopyOrMoveOperationType
+const (
+	MoveOriginal CopyOrMoveOperationType = iota
+	CopyOriginal
+	CopyOriginalPreserveModTime
+)
+
+// OnExistAction represents possible actions the API should take,
+// when a request tries to create a path that already exists.
+type OnExistAction string
+
+// possible values for OnExistAction
+const (
+	// IgnoreOnExist instructs the API not to execute
+	// the request in case of a conflict, but to return an error.
+	IgnoreOnExist OnExistAction = "ignore"
+	// AutoNameOnExist instructs the API to automatically rename
+	// any conflicting request-objects.
+	AutoNameOnExist OnExistAction = "autoname"
+	// OverwriteOnExist instructs the API to overwrite any conflicting files.
+	// This can only be used, if the request operates on files directly.
+	// (For example when moving/copying a file.)
+	// For most requests this action will simply be ignored.
+	OverwriteOnExist OnExistAction = "overwrite"
+)
+
+// shouldRetry returns a boolean as to whether this resp and err deserve to be retried.
+// It tries to expire/invalidate the token, if necessary.
+// It returns the err as a convenience.
+func (f *Fs) shouldRetry(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	if fserrors.ContextError(ctx, &err) {
+		return false, err
+	}
+	if resp != nil && (resp.StatusCode == 401 || isHTTPError(err, 401)) && len(resp.Header["Www-Authenticate"]) > 0 {
+		fs.Debugf(f, "Token might be invalid: %v", err)
+		if f.tokenRenewer != nil {
+			iErr := f.tokenRenewer.Expire()
+			if iErr == nil {
+				return true, err
+			}
+		}
+	}
+	return fserrors.ShouldRetry(err) || fserrors.ShouldRetryHTTP(resp, retryErrorCodes), err
+}
+
+// resolvePath resolves the given (relative) path and
+// returns a path suitable for API-calls.
+// This will consider the root-path of the fs and any needed prefixes.
+//
+// Any relative paths passed to functions that access these paths should
+// be resolved with this first!
+func (f *Fs) resolvePath(objectPath string) string {
+	resolved := path.Join(f.opt.RootPrefix, f.root, f.opt.Enc.FromStandardPath(objectPath))
+	if !f.opt.UnicodeNormalizationDisabled {
```

Comment:
I had to dig around in some team-discussions to remember the purpose of this one. :)
Apparently I originally planned to discuss this when opening the PR, but I forgot.

The unicode-normalization is required so that `fs/sync/sync_text.go:TestSyncUTFNorm` passes as intended.
It expects the two filenames to be treated equally, but HiDrive does not: We store filenames as raw-byte sequences and do not care about the encoding. Without the code above performing normalization on remote accesses, `rclone sync` will therefore fail to find the desired file, *because* it normalizes paths.

We decided that there are valid use cases for both accesses with and without normalization, so I made normalization optional.

Reply:
> In Update you are supposed to be updating an existing object, not creating a new one.

You are right, `overwriteFile` can create a new file, if it does not exist yet, as it's documentation indicates.
Maybe a direct call to `patchFile` with offset 0 would be more appropriate; it will fail if the file does not exist yet.

I'll need to think a bit about the rest of your findings.

Reply:
You're right, actually _none_ of the other remotes uses `src.Remote()` in the `Update` method. My bad!

So from what I understand there are 2 things here:
1. `Update` uses the wrong path, if it is supposed to update an existing object.
2. `Update` is able to create a new object, which it is not supposed to do.

Your simple suggestion from above completely fixes 1. which is great, but `Update` is still technically able to create new objects. This patch would solve the 2. problem:
```diff
--- a/backend/hidrive/hidrive.go
+++ b/backend/hidrive/hidrive.go
@@@ -984,7 -984,7 +984,13 @@@ func (o *Object) Update(ctx context.Con
                // Metadata should be updated even if the upload fails.
                info, metaErr = o.fs.fetchMetadataForPath(ctx, resolvedPath, api.HiDriveObjectWithMetadataFields)
        } else {
--              info, err = o.fs.overwriteFile(ctx, resolvedPath, cachedReader(in), modTime)
++              // Single-part upload with patchFile;
++              // it guarantees not to create a new object,
++              // but to overwrite the existing one.
++              err = o.fs.patchFile(ctx, resolvedPath, cachedReader(in), 0, time.Time{})
++              if err == nil {
++                      info, err = o.fs.resizeFile(ctx, resolvedPath, uint64(src.Size()), modTime)
++              }
                metaErr = err
        }
```
But it is a bit harder to understand from a first glance. Without your patch this would then lead to `TestSyncUTFNorm` not being able to access the file, instead of creating a new one.

Of course your patch makes this 2. problem more or less obsolete, so I am wondering if you like my above patch or would prefer to keep it how it was previously?
The only situation I see where this makes a difference is, if someone deletes the remote file before rclone can `Update` its contents: With this additional patch the operation will fail then, while without it, a new file with the same name and location as the previous one would be silently created. I guess that would then be a question of who has priority: rclone or anyone else using the remote.

Reply:
Also your patch makes my reasoning for introducing unicode normalization in the backend obsolete, since `TestSyncUTFNorm` passes now without it.

Though, from what I understand you would still like to keep it around (but to copy the existing flag from the local backend)?
Don't misunderstand me, I'm not opposed to keeping normalization around; I'm just wondering what your reasoning is.

Reply:
`Update` should try to update an existing file. The real purpose of this is for backends which can have duplicate file names (like Google Drive). It also helps with backends that keep an object history (eg revisions). Rclone will try to call `Update` if possible before calling `Put`.

Since your backend can't have duplicate file names, just using `o.remote` is probably enough. There may be reasons why you'd prefer to use patchfile - you'd be the expert on that.

So I'd say unless patchfile preserves some interesting metadata you'd lose otherwise (eg file revisions or something like that) then you should probably just upload it with the new name.